### DISCUSSION TITLE
How to build locally & with caching?

### DISCUSSION THREAD

**Discussion by takluyver** [mesonbuild/meson-python#644] [Q&A] [answered]
Is there a recommended way to quickly build a package locally, as fast as possible, for testing changes? With setuptools, I've done something like this:

```shell
python setup.py build --build-lib build/for-test
PYTHONPATH=build/for-test pytest
```

Using `pip install --no-build-isolation .` takes 4-5 seconds even with no changes, which is OK if there's nothing better, but it seems like it should be possible to go faster than that. Using the `meson` command can be much quicker, but `meson compile` doesn't copy the `.py` files into its build folder, and `meson install` wants to create a `site-packages/python3.12` style layout, which in this case isn't what I want.

Editable installs work nicely if you're only changing pure Python code, but I'd like to have a solution for changing Cython code with a fast build & test loop as well.

**Comment by rgommers** [comment, marked as answer]
Hey @takluyver, good question. A few options:

- Editable installs do work, and are automatically rebuilding also Cython  (and C, etc.) code if and only if source files are changed.
- A manual `meson compile` && `meson install` does work without touching site-packages if you configured initially with `meson setup builddir --prefix=$PWD/build-install` or something like that.
- https://github.com/scientific-python/spin/ UX provides a nice UX for essentially the manual steps above plus setting PYTHONPATH etc.

I'd recommend either using editable installs directly, or `spin`. Both should do what you want.

**Reply by takluyver** [reply]
Thanks @rgommers ! I've taken `spin` for a spin, it looks promising so far. :+1:
